### PR TITLE
Encoding fixes

### PR DESCRIPTION
**Pull request by eregon**
I think `encodeRope` should never be passed a binary encoding, otherwise it encodes the j.l.String in  ISO-8859-1 bytes, which is incorrect if any of them is higher than 128.
The problem is if we have just a j.l.String, we no longer have the raw bytes, and we need to know where does the String comes from to reverse the process bytes->j.l.String.

I was thinking to add an assertion in encodeRope that a binary encoding is not passed, as it re-encodes chars to ISO-8859-1 which seems to make no sense as it's very unlikely to be the original encoding.
But there is one usage in SymbolParseNode and I am not sure whether it's fine to use ASCIIEncoding there.
In any case, I wrote a spec showing it is not working as expected in SymbolParseNode (somehow it believes the Symbol is US-ASCII and 7-bit).

### CODE REVIEW

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 1]
```diff
@@ -1120,7 +1120,7 @@ public int hash(DynamicObject string,
 
         @Specialization
         public DynamicObject initializeJavaString(DynamicObject self, String from) {
-            StringOperations.setRope(self, StringOperations.encodeRope(from, ASCIIEncoding.INSTANCE));
+            StringOperations.setRope(self, StringOperations.encodeRope(from, UTF8Encoding.INSTANCE));
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: UTF-8 is safer for sure. But I think this starts getting into questions like is a j.l.String an "external" string? Should it be subject to the external encoding then?

File: spec/tags/language/symbol_tags.txt [hunk 2]
```diff
@@ -0,0 +1,2 @@
+slow:A Symbol literal inherits the encoding of the magic comment and can have a binary encoding
+fails:A Symbol literal inherits the encoding of the magic comment and can have a binary encoding
```
Comment (nirvdrum) [inline spec/tags/language/symbol_tags.txt]: I looked into this a bit. For some reason, the lexer thinks the code range should be `CR_7BIT`. Debugging that part is going to take more effort. `SymbolParseNode` then goes on to encode the j.l.String to a rope with `USASCIIEncoding.INSTANCE`. That process converts that "bad" bytes to '?' and so the resulting rope is well-formed. It just no longer represents the same set of characters.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: a j.l.String which comes from interop could be from anywhere (e.g. a literal String in JS code). I think j.l.String are meant to be textual and non-binary in most cases. So I think UTF-8 is a good choice here for compatibility with other Ruby Strings.

File: spec/tags/language/symbol_tags.txt (hunk 2 above)
Comment (eregon) [reply spec/tags/language/symbol_tags.txt]: Yes, the lexer definitely gets this wrong for some reason. And the encoding of the Symbol should be binary.

**Comment by eregon** [comment]
Actually I decided to do this the proper way by requiring an Encoding when passing a javaString to String.new, see cdec141.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: That doesn't work for Strings coming from C which should be tagged as binary, so String.new(javaString) now requires an Encoding to be passed.

Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: Right. I think we want to delineate between strings from other languages and strings from NFI/Sulong. We haven't done such a good job of this before, but we should be using `Encoding.default_external` when making system calls in most (all?) cases.

Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: For info, is `Encoding.default_external` the same as `Encoding.find('locale')` ?

**Pull request merged**